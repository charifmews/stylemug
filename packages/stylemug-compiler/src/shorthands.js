export function expandShorthand(key, value) {
  const values = value.split(' ');

  const def = Shorthands[key];
  if (!def) {
    return null;
  }

  const defaults = Object.values(def);
  return Object.keys(def).reduce((obj, key, i) => {
    obj[key] = values[i] || defaults[i];
    return obj;
  }, {});
}
