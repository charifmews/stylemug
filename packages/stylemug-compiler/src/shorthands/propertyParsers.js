const isUnit = (val) =>
  val.match(
    /(em|ex|ch|rem|vw|vh|vmin|vmax|cm|mm|q|in|pt|pc|px|dpi|dpcm|dppx|%|auto)$/i
  ) !== null ||
  val.match(/^(calc\()/i) !== null ||
  val.match(/^(var\()/i) !== null;

const createParser = (parseFn) => (resolve) => (value) => {
  const values = [];
  let temp = '';
  let bracketsCount = 0;

  const trim = value.toString().trim();
  const len = trim.length;

  for (let i = 0; i < len; i += 1) {
    const char = trim.charAt(i);
    const inBrackets = bracketsCount > 0;

    if (!inBrackets && char === ' ') {
      values.push(temp);
      temp = '';
    } else {
      temp += char;
    }

    // Detect whether we are in brackets or not.
    bracketsCount +=
      {
        '(': 1,
        ')': -1,
      }[char] || 0;
  }

  if (temp) {
    values.push(temp);
  }

  return parseFn(values, resolve);
};

export const borderParser = createParser((values, resolve) => {
  const BORDER_STYLE = /^(dashed|dotted|double|groove|hidden|inset|none|outset|ridge|solid)$/i;
  const BORDER_WIDTH = /^(thick|medium|think)$/i;

  const longhands = {};

  values.forEach((value) => {
    if (value.match(BORDER_STYLE) !== null) {
      longhands[resolve('Style')] = value;
    } else if (
      isUnit(value) ||
      value.match(BORDER_WIDTH) !== null ||
      value === '0'
    ) {
      longhands[resolve('Width')] = value;
    } else {
      longhands[resolve('Color')] = value;
    }
  });

  return longhands;
});

export const circularParser = createParser((values, resolve) => {
  const longhands = {};

  longhands[resolve('Top')] = values[0];
  longhands[resolve('Right')] = values[1] || values[0];
  longhands[resolve('Bottom')] = values[2] || values[0];
  longhands[resolve('Left')] = values[3] || values[1] || values[0];

  return longhands;
});

export const flexParser = createParser((values) => {
  const longhands = {};

  values.forEach((value) => {
    if (isUnit(value)) {
      longhands.flexBasis = value;
    } else {
      if (longhands.flexGrow) {
        longhands.flexShrink = value;
      } else {
        longhands.flexGrow = value;
      }
    }
  });

  return longhands;
});
