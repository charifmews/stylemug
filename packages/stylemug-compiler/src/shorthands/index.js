import { circularParser, borderParser, flexParser } from './propertyParsers';

const Expanders = {
  padding: circularParser((key) => `padding${key}`),
  margin: circularParser((key) => `margin${key}`),
  borderWidth: circularParser((key) => `border${key}Width`),
  borderColor: circularParser((key) => `border${key}Color`),
  borderStyle: circularParser((key) => `border${key}Style`),
  borderLeft: borderParser((key) => `borderLeft${key}`),
  borderTop: borderParser((key) => `borderTop${key}`),
  borderRight: borderParser((key) => `borderRight${key}`),
  borderBottom: borderParser((key) => `borderBottom${key}`),
  outline: borderParser((key) => `outline${key}`),
  flex: flexParser(),
};

function expandProperty(property, value) {
  if (property === 'border') {
    const longhands = borderParser((key) => `border${key}`)(value);

    var result = {};
    for (let property in longhands) {
      Object.assign(result, expandProperty(property, longhands[property]));
    }

    return result;
  }

  if (Expanders[property]) {
    const parse = Expanders[property];
    return parse(value);
  }
}

export function extractShorthands(rules) {
  const priority = [
    'borderLeft',
    'borderRight',
    'borderBottom',
    'borderTop',
    'borderWidth',
    'borderStyle',
    'borderColor',
  ];

  const sortedKeys = Object.keys(rules)
    .sort((a, b) =>
      priority.indexOf(a) && priority.indexOf(b)
        ? priority.indexOf(a) > priority.indexOf(b)
        : a > b || -1
    )
    .reverse();

  for (let key of sortedKeys) {
    const value = rules[key];

    if (
      value === null ||
      (typeof value !== 'string' && typeof value !== 'number')
    ) {
      continue;
    }

    // Figure out if we have an expansion for a given rule.
    const expansion = expandProperty(key, value);
    if (!expansion) {
      continue;
    }

    // If we have a rule defined for our expansion, use that value
    // instead.
    // {
    //   margin: '1px 2px', // expands
    //   marginTop: '5px', // use defined value instead.
    // }
    for (let key in expansion) {
      if (rules.hasOwnProperty(key)) {
        expansion[key] = rules[key];
      }
    }

    Object.assign(rules, expansion);

    // Remove the key, it has been replaced with expanded rules.
    delete rules[key];
  }

  return rules;
}
