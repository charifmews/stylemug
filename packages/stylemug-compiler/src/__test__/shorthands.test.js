import { extractShorthands } from '../shorthands';

describe('merge', () => {
  it.each([
    ['padding', '1px 2px 3px 4px'],
    ['padding', '1px 2px 3px'],
    ['padding', '1px 2px'],
    ['padding', '1px'],
    ['padding', 1],
    ['margin', '1px 2px 3px 4px'],
    ['margin', '1px 2px 3px'],
    ['margin', '1px 2px'],
    ['margin', '1px'],
    ['margin', 1],
  ])('expand circular %s with "%s" in given stylesheet', (name, value) => {
    const style = extractShorthands({
      [name]: value,
    });
    expect(style).toMatchSnapshot();
  });

  it.each([
    ['padding', { padding: '1px 2px' }, { paddingTop: '10px' }],
    ['margin', { margin: '1px 2px' }, { marginTop: '10px' }],
  ])('should override defined shorthands for %s', (name, prop1, prop2) => {
    const style = extractShorthands({ ...prop1, ...prop2 });
    expect(style).toMatchSnapshot();

    const styleReversed = extractShorthands({ ...prop2, ...prop1 });
    expect(styleReversed).toMatchSnapshot();

    expect(style).toEqual(styleReversed);
  });

  it('should expand flex', () => {
    expect(
      extractShorthands({
        flex: '0 1 auto',
      })
    ).toMatchSnapshot();

    expect(
      extractShorthands({
        flex: '1 1',
      })
    ).toMatchSnapshot();
  });

  it('should merge properties', () => {
    const style = extractShorthands({
      padding: '1px 2px',
      paddingLeft: '10px',
      paddingTop: 20,
    });
    expect(style).toMatchSnapshot();
  });

  it('should merge border properties', () => {
    const style = extractShorthands({
      border: '1px solid red',
      borderLeft: '10px dotted yellow',
      borderWidth: '20px 30px',
      borderLeftWidth: 5,
    });
    expect(style).toMatchSnapshot();
  });

  it('should merge flex', () => {
    const style = extractShorthands({
      flex: '0 1 auto',
      flexGrow: 1,
    });
    expect(style).toMatchSnapshot();
  });

  it('should preserve non shorthand properties', () => {
    const style = extractShorthands({
      padding: 1,
      color: 'red',
      foo: 'bar',
    });
    expect(style.color).toBe('red');
    expect(style.foo).toBe('bar');
  });

  it('should parse var', () => {
    const style = extractShorthands({
      padding: '10px var(--foo)',
    });
    expect(style).toMatchSnapshot();
  });

  it('should parse calc', () => {
    const style = extractShorthands({
      padding: 'calc(100% - 1px)',
    });
    expect(style).toMatchSnapshot();
  });

  it('should parse a mix of var and calc', () => {
    const style = extractShorthands({
      margin: 'calc(100% - var(--foo)) 10px',
    });
    expect(style).toMatchSnapshot();
  });
});
