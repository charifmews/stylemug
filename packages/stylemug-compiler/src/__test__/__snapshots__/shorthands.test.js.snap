// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge expand circular margin with "1" in given stylesheet 1`] = `
Object {
  "marginBottom": "1",
  "marginLeft": "1",
  "marginRight": "1",
  "marginTop": "1",
}
`;

exports[`merge expand circular margin with "1px 2px 3px 4px" in given stylesheet 1`] = `
Object {
  "marginBottom": "3px",
  "marginLeft": "4px",
  "marginRight": "2px",
  "marginTop": "1px",
}
`;

exports[`merge expand circular margin with "1px 2px 3px" in given stylesheet 1`] = `
Object {
  "marginBottom": "3px",
  "marginLeft": "2px",
  "marginRight": "2px",
  "marginTop": "1px",
}
`;

exports[`merge expand circular margin with "1px 2px" in given stylesheet 1`] = `
Object {
  "marginBottom": "1px",
  "marginLeft": "2px",
  "marginRight": "2px",
  "marginTop": "1px",
}
`;

exports[`merge expand circular margin with "1px" in given stylesheet 1`] = `
Object {
  "marginBottom": "1px",
  "marginLeft": "1px",
  "marginRight": "1px",
  "marginTop": "1px",
}
`;

exports[`merge expand circular padding with "1" in given stylesheet 1`] = `
Object {
  "paddingBottom": "1",
  "paddingLeft": "1",
  "paddingRight": "1",
  "paddingTop": "1",
}
`;

exports[`merge expand circular padding with "1px 2px 3px 4px" in given stylesheet 1`] = `
Object {
  "paddingBottom": "3px",
  "paddingLeft": "4px",
  "paddingRight": "2px",
  "paddingTop": "1px",
}
`;

exports[`merge expand circular padding with "1px 2px 3px" in given stylesheet 1`] = `
Object {
  "paddingBottom": "3px",
  "paddingLeft": "2px",
  "paddingRight": "2px",
  "paddingTop": "1px",
}
`;

exports[`merge expand circular padding with "1px 2px" in given stylesheet 1`] = `
Object {
  "paddingBottom": "1px",
  "paddingLeft": "2px",
  "paddingRight": "2px",
  "paddingTop": "1px",
}
`;

exports[`merge expand circular padding with "1px" in given stylesheet 1`] = `
Object {
  "paddingBottom": "1px",
  "paddingLeft": "1px",
  "paddingRight": "1px",
  "paddingTop": "1px",
}
`;

exports[`merge should expand flex 1`] = `
Object {
  "flexBasis": "auto",
  "flexGrow": "0",
  "flexShrink": "1",
}
`;

exports[`merge should expand flex 2`] = `
Object {
  "flexGrow": "1",
  "flexShrink": "1",
}
`;

exports[`merge should merge border properties 1`] = `
Object {
  "borderBottomColor": "red",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "20px",
  "borderLeftColor": "yellow",
  "borderLeftStyle": "dotted",
  "borderLeftWidth": 5,
  "borderRightColor": "red",
  "borderRightStyle": "solid",
  "borderRightWidth": "30px",
  "borderTopColor": "red",
  "borderTopStyle": "solid",
  "borderTopWidth": "20px",
}
`;

exports[`merge should merge flex 1`] = `
Object {
  "flexBasis": "auto",
  "flexGrow": 1,
  "flexShrink": "1",
}
`;

exports[`merge should merge properties 1`] = `
Object {
  "paddingBottom": "1px",
  "paddingLeft": "10px",
  "paddingRight": "2px",
  "paddingTop": 20,
}
`;

exports[`merge should override defined shorthands for margin 1`] = `
Object {
  "marginBottom": "1px",
  "marginLeft": "2px",
  "marginRight": "2px",
  "marginTop": "10px",
}
`;

exports[`merge should override defined shorthands for margin 2`] = `
Object {
  "marginBottom": "1px",
  "marginLeft": "2px",
  "marginRight": "2px",
  "marginTop": "10px",
}
`;

exports[`merge should override defined shorthands for padding 1`] = `
Object {
  "paddingBottom": "1px",
  "paddingLeft": "2px",
  "paddingRight": "2px",
  "paddingTop": "10px",
}
`;

exports[`merge should override defined shorthands for padding 2`] = `
Object {
  "paddingBottom": "1px",
  "paddingLeft": "2px",
  "paddingRight": "2px",
  "paddingTop": "10px",
}
`;

exports[`merge should parse a mix of var and calc 1`] = `
Object {
  "marginBottom": "calc(100% - var(--foo))",
  "marginLeft": "10px",
  "marginRight": "10px",
  "marginTop": "calc(100% - var(--foo))",
}
`;

exports[`merge should parse calc 1`] = `
Object {
  "paddingBottom": "calc(100% - 1px)",
  "paddingLeft": "calc(100% - 1px)",
  "paddingRight": "calc(100% - 1px)",
  "paddingTop": "calc(100% - 1px)",
}
`;

exports[`merge should parse var 1`] = `
Object {
  "paddingBottom": "10px",
  "paddingLeft": "var(--foo)",
  "paddingRight": "var(--foo)",
  "paddingTop": "10px",
}
`;
